src/plot_api/plot_api.js

```diff
@@ -2579,61 +2601,60 @@ Plotly.animate = function(gd, frameOrGroupNameOrFrameList, animationOpts) {
                 trans._lastFrameAt = Date.now();
                 trans._timeToNext = newFrame.frameOpts.duration;
 
+                // This is simply called and it's left to .transition to decide how to manage
+                // interrupting current transitions. That means we don't need to worry about
+                // how it resolves or what happens after this:
                 Plots.transition(gd,
                     newFrame.frame.data,
                     newFrame.frame.layout,
                     newFrame.frame.traces,
                     newFrame.frameOpts,
                     newFrame.transitionOpts
-                ).then(function() {
-                    if(trans._frameQueue.length === 0) {
-                        gd.emit('plotly_animated');
-                        if(trans._currentFrame && trans._currentFrame.onComplete) {
-                            trans._currentFrame.onComplete();
-                            trans._currentFrame = null;
-                        }
-                    }
-                });
-            }
-
-            if(trans._frameQueue.length === 0) {
+                );
+            } else {
+                // If there are no more frames, then stop the RAF loop:
                 stopAnimationLoop();
             }
         }
 
         function beginAnimationLoop() {
             gd.emit('plotly_animating');
 
-            // If no timer is running, then set last frame = long ago:
+            // If no timer is running, then set last frame = long ago so that the next
+            // frame is immediately transitioned:
             trans._lastFrameAt = -Infinity;
             trans._timeToNext = 0;
             trans._runningTransitions = 0;
             trans._currentFrame = null;
 
             var doFrame = function() {
-                // Check if we need to pop a frame:
+                // This *must* be requested before nextFrame since nextFrame may decide
+                // to cancel it if there's nothing more to animated:
+                trans._animationRaf = window.requestAnimationFrame(doFrame);
+
+                // Check if we're ready for a new frame:
                 if(Date.now() - trans._lastFrameAt > trans._timeToNext) {
                     nextFrame();
                 }
-
-                trans._animationRaf = window.requestAnimationFrame(doFrame);
```

Comment:
@etpinard This was the source of some other weirdness. The raf was getting set *after* `nextFrame()` was called so that cancelAnimationFrame had no effect.